src/main/java/org/truffleruby/core/array/ArrayNodes.java

```diff
@@ -1922,35 +1919,69 @@ public DynamicObject sortVeryShort(VirtualFrame frame, DynamicObject array, NotP
+        @Specialization(guards = { "!isEmptyArray(array)", "!isSmall(array)", "isIntArray(array)" },
+                        assumptions = "getContext().getCoreMethods().fixnumCmpAssumption")
+        public Object sortIntArray(DynamicObject array, NotProvided block) {
+            final int size = getSize(array);
+            int[] copy = ((int[]) getStore(array)).clone();
+            doSort(copy, size);
+            return createArray(copy, size);
         }
 
-        @Specialization(guards = { "!isEmptyArray(array)", "isObjectArray(array)" })
-        public Object sortObjectWithBlock(DynamicObject array, DynamicObject block) {
+        @Specialization(guards = { "!isEmptyArray(array)", "!isSmall(array)", "isLongArray(array)" },
+                        assumptions = "getContext().getCoreMethods().fixnumCmpAssumption")
+        public Object sortLongArray(DynamicObject array, NotProvided block) {
             final int size = getSize(array);
-            Object[] copy = ((Object[]) getStore(array)).clone();
-            doSort(copy, size, block);
+            long[] copy = ((long[]) getStore(array)).clone();
+            doSort(copy, size);
             return createArray(copy, size);
         }
 
-        @TruffleBoundary
-        private void doSort(Object[] copy, int size, DynamicObject block) {
-            Arrays.sort(copy, 0, size, (a, b) -> castSortValue(yieldNode.dispatch(block, a, b)));
+        @Specialization(guards = { "!isEmptyArray(array)", "!isSmall(array)", "isDoubleArray(array)" },
+                        assumptions = "getContext().getCoreMethods().floatCmpAssumption")
+        public Object sortDoubleArray(DynamicObject array, NotProvided block) {
+            final int size = getSize(array);
+            double[] copy = ((double[]) getStore(array)).clone();
+            doSort(copy, size);
+            return createArray(copy, size);
+        }
+
+        @Specialization(guards = { "!isEmptyArray(array)", "isObjectArray(array)" })
+        public Object sortObjectArrayWithoutBlock(VirtualFrame frame, DynamicObject array, NotProvided block,
+                                                  @Cached("new()") SnippetNode snippetNode) {
+            return sortArrayInRuby(frame, snippetNode);
         }
 
         @Specialization(guards = { "!isEmptyArray(array)", "!isObjectArray(array)" })
-        public Object sortWithBlock(VirtualFrame frame, DynamicObject array, DynamicObject block,
-                @Cached("new()") SnippetNode snippet) {
-            return snippet.execute(frame,
-                    "sorted = dup; Truffle.privately { sorted.mergesort_block!(block) }; sorted",
-                    "block", block);
+        public Object sortPrimitiveArrayFallback(VirtualFrame frame, DynamicObject array, NotProvided block,
+                                                 @Cached("new()") SnippetNode snippetNode) {
+            return sortArrayInRuby(frame, snippetNode);
```

Comment:
I think it would be nicer to have this specializations just below `sortDoubleArray`, so the fallback is directly below and related specializations stay together.